File: src/main/scala/io/shinto/amaterasu/execution/actions/Action.scala
package io.shinto.amaterasu.execution.actions

import io.shinto.amaterasu.Logging
import io.shinto.amaterasu.dataObjects.ActionData
import io.shinto.amaterasu.enums.ActionStatus
import org.apache.curator.framework.CuratorFramework

trait Action extends Logging {

  // this is the znode path for the action
  var actionPath: String = _
  var actionId: String = _

  var data: ActionData = null
  var client: CuratorFramework = null

  def execute(): Unit

  def handleFailure(message: String): String

  /**
    * The announceStart register the beginning of the of the task with ZooKeper
    */
  def announceStart: Unit = {

    log.debug(s"Starting action ${data.name} of type ${data.actionType}")
    client.setData().forPath(actionPath, ActionStatus.started.toString.getBytes)

  }

  def announceQueued: Unit = {

    log.debug(s"Action ${data.name} of type ${data.actionType} is queued for execution")
    client.setData().forPath(actionPath, ActionStatus.queued.toString.getBytes)

  }

  def announceComplete: Unit = {

    log.debug(s"Action ${data.name} of type ${data.actionType} completed")
    client.setData().forPath(actionPath, ActionStatus.complete.toString.getBytes)

  }

  protected def announceFailure(): Unit = {}

}

File: leader/src/main/kotlin/org/apache/amaterasu/leader/execution/Action.kt
package org.apache.amaterasu.leader.execution


import org.apache.amaterasu.common.dataobjects.ActionData
import org.apache.amaterasu.common.configuration.enums.ActionStatus
import org.apache.amaterasu.common.logging.Logging
import org.apache.curator.framework.CuratorFramework

/**
 * Created by Eran Bartenstein on 19/10/18.
 */
abstract class Action() : Logging {
    lateinit var actionPath: String
    lateinit var actionId: String
    lateinit var client: CuratorFramework
    lateinit var data: ActionData
    abstract fun execute()
    abstract fun handleFailure(message: String) : String

    fun announceStart() {
        log().debug("Starting action ${data.name} of group ${data.groupId} and type ${data.typeId}")
        val startedAction: String = ActionStatus.started.value
        client.setData().forPath(actionPath, startedAction.toByteArray())
        data.status = ActionStatus.started
    }
}